core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/fuse/FuseCrossJoinedGlobalAggregations.java

```diff
@@ -0,0 +1,138 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule.fuse;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.tree.SymbolReference;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.SystemSessionProperties.isFuseSubPlanEnabled;
+import static io.trino.sql.planner.iterative.rule.fuse.PlanNodeFuser.outputsSameAsSource;
+import static io.trino.sql.planner.plan.Patterns.join;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Simplified "Computation Reuse via Fusion in Amazon Athena" paper JoinOnKeys rule
+ * that works only on a single cross join over global aggregations over the same table.
+ * It transforms
+ * <pre>
+ *   cross join
+ *     aggregation global
+ *       agg_left <- aggLeft()
+ *         filter(filter_left)
+ *           table scan(table)
+ *     aggregation global
+ *       agg_right <- aggRight()
+ *         filter(filter_right)
+ *           table scan(table)
+ * </pre>
+ * into:
+ * <pre>
+ *   aggregation global
+ *       agg_left <- aggLeft(mask = agg_mask_left)
+ *       agg_right <- aggRight(mask = agg_mask_right)
+ *     project(agg_mask_left = filter_left, agg_mask_left = filter_right)
+ *       filter(filter_left or filter_right)
+ *         table scan
+ * </pre>
+ * <p>
+ * This makes the plan to read the source table only once,
+ * resulting in the potentially significant performance improvement.
+ */
+public class FuseCrossJoinedGlobalAggregations
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join()
+            .matching(JoinNode::isCrossJoin);
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isFuseSubPlanEnabled(session);
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode left = context.getLookup().resolve(node.getLeft());
+        PlanNode right = context.getLookup().resolve(node.getRight());
+
+        while (left instanceof ProjectNode && outputsSameAsSource((ProjectNode) left)) {
```

Comment:
I think these conditions are not required if this rule is fired after `RemoveRedundantIdentityProjections`

Reply:
there is an order of rule issue. this rule cannot be run after `RemoveRedundantIdentityProjections` because it has to be run before columnPruningOptimizer, PushPredicateIntoTableScan, AddExchanges.
I could add another `RemoveRedundantIdentityProjections` before this rule but this changes plans for some unrelated queries on values and I want this change to be as small as possible.

Reply:
I don't there's any harm in adding `RemoveRedundantIdentityProjections` and it simplifies the fuse rule since it only has to take care of the fusing logic and nothing else. WDYT?

Reply:
There are two things. One is, adding this `RemoveRedundantIdentityProjections` before `FuseCrossJoinedGlobalAggregations` changes plans in unrelated queries in a not-obvious way (it reorders the symbols in `VALUES` node) but also it runs `RemoveRedundantIdentityProjections` for the entire plan even if the `FuseCrossJoinedGlobalAggregations` won't fire, adding to the planning time.